### ISSUE TITLE
Add support for Deserialization only

### ISSUE THREAD

**Issue by feinstein** [google/built_value.dart#917]
I have some Value Objects that are representations from a Firebase Cloud Firestore database. Those VOs need to hold the document id they have on the database, so later I can update or delete them, using that `id`.

My VOs look like:

```dart
abstract class MyClass implements Built<MyClass, MyClassBuilder> {
  static Serializer<MyClass> get serializer => _$myClassSerializer;

  @BuiltValueField(serialize: false)
  String get id;
...
```

And I construct them with Firestore like:

```dart
return serializers.deserializeWith(
  MyClass.serializer,
  document.data()..putIfAbsent('id', () => document.id),
);
```

So upon deserialization I add the document's `id` into the serialized data so it can be easily passed to the VO.

The problem is `@BuiltValueField(serialize: false)` cancels serialization AND deserialization (silly me got things literal and thought it would only cancel the serialization itself, not the deserialization).

I know I can make it `@nullable` or add a default value, but this isn't a perfect solution as neither `null` or a default value are correct or acceptable for an `id` and can creep errors into the App.

Could we have a mechanism for selecting if we want `serialization`, `deserialization` or `both`? Maybe an extra enum field for configuring the `serialization`?

**Comment by davidmorgan** [comment]
I think the simplest thing would be to turn serialization for the field on, then remove the `id` key from the resulting map yourself.

It might be possible to make the generation configurable as you suggest; it'll be a while before I get around to doing another batch of features, though, as I'm waiting for null safety to be fully released.

**Comment by feinstein** [comment]
Do you mean setting `id` as `@nullable` and setting it to `null` before serializing? That's what I am doing right now. 

I just don't like because it opens the possibility for a `null` `id`, and adding extra checks on the private constructor is cumbersome as I have some 7 VOs as of now.

I understand this is not a priority, no problem, I appreciate the help :).

**Comment by davidmorgan** [comment]
No, I meant serialize with `id`, then remove `id` from the serialized data :) something like

```
return serializers.serializeWith(
  MyClass.serializer,
  myClassValue,
)..remove('id');
```

**Comment by feinstein** [comment]
Oh, good point!

I am rethinking the whole thing right now tbh, as I realized I need a `@nullable` `id` since VOs won't have an `id` until after they are added into the database.

That's what happens when we code until late lol.

**Comment by larssn** [comment]
@davidmorgan 
This WOULD be a great feature though, request we reopen it.

It would allow for contractually guaranteeing that certain fields CANNOT be updated by the client, and only the server, which is a strategy we heavily use.

And we've often had problems with a field accidentally getting serialized into the database, because the entire object was serialized (with `serializers.serializeWith(serializer, this)` as all our built value classes have).

This would be a great way to automatically prevent that.

**Comment by davidmorgan** [comment]
Thanks @larssn.

Makes sense.

Happy to reopen, please anyone who would also use this feature add your +1.

At first glance though it looks unlikely to happen given how much (little) bandwidth I have for improvements and the amount of demand for it.

**Comment by feinstein** [comment]
Sorry to say it, but I switched to `freezed` many years ago, so I don't use this lib anymore.

**Comment by larssn** [comment]
@davidmorgan Thanks for reopening.

> At first glance though it looks unlikely to happen given how much (little) bandwidth I have for improvements and the amount of demand for it.

Thats totally fair, but maybe we should discuss how the API should be, in case someone (maybe me) (also have 0 time) should do a PR for it.

I thought about simply having a `deserialize` field in `BuiltValueField`, but that would basically be a breaking change, as it would implicitly influence the `serialize` field.

So maybe a `allowDeserialization` field instead, so that if this field is true and `serialize` is false, then the associated serializer's `deserialize` method, won't be touched.

What do you think? The task in itself should be relatively simpel.

**Comment by davidmorgan** [comment]
Yeah, the awkwardness of the API is the main thing that makes the change hard.

How about `onlyDeserialize`:

```
Pass [onlyDeserialize] to ignore the value of [serialize] and cause the field to be
never written on serialization, but deserialized if present. This creates a field that
can only ever be written on the wire due to some other system, for example if
the server uses a different serializer implementation.
```

**Comment by larssn** [comment]
I can get behind that idea.

Maybe we should force it, so only one field can be filled out? Just to avoid confusing settings.
We could have a sanity check, if both this field and `serialize` is set. Like:

```dart
assert(serialize != null && onlyDeserialize != null,
'Both serialize and onlyDeserialize had values other than null! This is not allowed, and only one of them can have a value'
);
```

**Comment by davidmorgan** [comment]
Yes, I think that works.
